circuits/lib/authV2.circom

```diff
@@ -0,0 +1,113 @@
+    signal input globalSmtMtp[onChainLevels];
+    signal input globalSmtMtpAuxHi;
+    signal input globalSmtMtpAuxHv;
+    signal input globalSmtMtpNoAux;
+
+    // userID output signal will be assigned with user profile hash(UserID, nonce),
+    // unless nonce == 0, in which case userID will be assigned with userGenesisID
+    signal output userID;
+
+
+
+    /* id ownership check */
+    component checkIdOwnership = IdOwnership(IdOwnershipLevels);
+
+    checkIdOwnership.userClaimsTreeRoot <== userClaimsTreeRoot;
+    for (var i=0; i<IdOwnershipLevels; i++) { checkIdOwnership.userAuthClaimMtp[i] <== userAuthClaimMtp[i]; }
+    for (var i=0; i<8; i++) { checkIdOwnership.userAuthClaim[i] <== userAuthClaim[i]; }
+
+    checkIdOwnership.userRevTreeRoot <== userRevTreeRoot;
+    for (var i=0; i<IdOwnershipLevels; i++) { checkIdOwnership.userAuthClaimNonRevMtp[i] <== userAuthClaimNonRevMtp[i]; }
+    checkIdOwnership.userAuthClaimNonRevMtpNoAux <== userAuthClaimNonRevMtpNoAux;
+    checkIdOwnership.userAuthClaimNonRevMtpAuxHv <== userAuthClaimNonRevMtpAuxHv;
+    checkIdOwnership.userAuthClaimNonRevMtpAuxHi <== userAuthClaimNonRevMtpAuxHi;
+
+    checkIdOwnership.userRootsTreeRoot <== userRootsTreeRoot;
+
+    checkIdOwnership.challenge <== challenge;
+    checkIdOwnership.challengeSignatureR8x <== challengeSignatureR8x;
+    checkIdOwnership.challengeSignatureR8y <== challengeSignatureR8y;
+    checkIdOwnership.challengeSignatureS <== challengeSignatureS;
+
+    checkIdOwnership.userState <== userState;
+
+    /* Check on-chain SMT inclusion existence */
+    component cutId = cutId();
+    cutId.in <== userGenesisID;
+
+    component cutState = cutState();
+    cutState.in <== userState;
+
+    component isStateGenesis = IsEqual();
+    isStateGenesis.in[0] <== cutId.out;
+    isStateGenesis.in[1] <== cutState.out;
+
+    component userGenesisIDhash = Poseidon(1);
+    userGenesisIDhash.inputs[0] <== userGenesisID;
+
+    component onChainSmtCheck = SMTVerifier(onChainLevels);
+    onChainSmtCheck.enabled <== 1;
+    onChainSmtCheck.fnc <== isStateGenesis.out; // non-inclusion in case if genesis state, otherwise inclusion
+    onChainSmtCheck.root <== globalSmtRoot;
+    for (var i=0; i<onChainLevels; i++) { onChainSmtCheck.siblings[i] <== globalSmtMtp[i]; }
+    onChainSmtCheck.oldKey <== globalSmtMtpAuxHi;
+    onChainSmtCheck.oldValue <== globalSmtMtpAuxHv;
+    onChainSmtCheck.isOld0 <== globalSmtMtpNoAux;
+    onChainSmtCheck.key <== userGenesisIDhash.out;
+    onChainSmtCheck.value <== userState;
+
+    /* ProfileID calculation */
+    component calcProfile = ProfileID();
+    calcProfile.in <== userGenesisID;
+    calcProfile.nonce <== nonce;
+
+    component isSaltZero = IsZero();
+    isSaltZero.in <== nonce;
+
+    component selectProfile = Mux1();
+    selectProfile.s <== isSaltZero.out;
+    selectProfile.c[0] <== calcProfile.out;
+    selectProfile.c[1] <== userGenesisID;
```

Comment:
You duplicated the same code in the `verifyCredentialSubjectProfile` template. Can be separated to utils

Reply:
Move selecting profile to a separate template